File: benchmarks-QuTiP/multiplication_dense_dense.py
import scipy.sparse as sp
import benchmarkutils

name = "multiplication_dense_dense"

samples = 2
evals = 5
cutoffs = range(50, 601, 50)


def setup(N):
    op1 = sp.rand(N, N, 1., dtype=float) * 0.2j
    op2 = sp.rand(N, N, 1., dtype=float) * 0.1j
    return op1, op2


def f(op1, op2):
    return op1 * op2


print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    op1, op2 = setup(N)
    t = benchmarkutils.run_benchmark(f, op1, op2, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.save(name, results)

File: benchmarks-QuantumOptics.jl/multiplication_dense_dense.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

using Random; Random.seed!(0)

name = "multiplication_dense_dense"

samples = 3
evals = 10
cutoffs = [50:50:1000;]

function setup(N)
    b = GenericBasis(N)
    op1 = randoperator(b)
    op2 = randoperator(b)
    result = DenseOperator(b)
    op1, op2, result
end

function f(op1, op2, result)
    QuantumOpticsBase.gemm!(ComplexF64(1., 0.), op1, op2, ComplexF64(0., 0.), result)
end

println("Benchmarking: ", name)
print("Cutoff: ")
results = []
for N in cutoffs
    print(N, " ")
    op1, op2, result = setup(N)
    t = @belapsed f($op1, $op2, $result) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.save(name, results)
